mdata/reorder_buffer.go

```diff
@@ -30,14 +30,16 @@ func NewReorderBuffer(reorderWindow, interval uint32, allowUpdate bool) *Reorder
 
 // Add adds the point if it falls within the window.
 // it returns points that have been purged out of the buffer, as well as whether the add succeeded.
-func (rob *ReorderBuffer) Add(ts uint32, val float64) ([]schema.Point, error) {
+func (rob *ReorderBuffer) Add(ts uint32, val float64) (schema.Point, []schema.Point, error) {
 	ts = AggBoundary(ts, rob.interval)
 
+	newestTs := rob.buf[rob.newest].Ts
 	// out of order and too old
-	if rob.buf[rob.newest].Ts != 0 && ts <= rob.buf[rob.newest].Ts-(uint32(cap(rob.buf))*rob.interval) {
-		return nil, errors.ErrMetricTooOld
+	if newestTs != 0 && ts < newestTs && ts <= newestTs-(uint32(cap(rob.buf))*rob.interval) {
```

Comment:
this still results in errors.ErrMetricTooOld.
e.g. with a window of 10, after having inserted a point with ts = 5,
if you try to add a point with ts 4, this error will trigger.

this can be demonstrated with a modified version of your test:
```
func TestROBAvoidUnderflow(t *testing.T) {
	var results []schema.Point
	windowSize := 10
	buf := NewReorderBuffer(uint32(windowSize), 1, false)
	data := make([]schema.Point, 5)
	for i := 0; i <= 4; i++ {
		data[i] = schema.Point{Ts: uint32(i + 1), Val: float64(i + 1000)}
	}
	for i := len(data) - 1; i >= 0; i-- {
		pt, flushed, err := buf.Add(data[i].Ts, data[i].Val)
		fmt.Println("adding", data[i].Ts, "results in", pt, flushed, err)
		if pt.Ts != 0 {
			results = append(results, pt)
			results = append(results, flushed...)
		}
	}

	expectedLen := len(data) - windowSize
	if len(results) != expectedLen {
		t.Fatalf("Expected %d results, got %d", expectedLen, len(results))
	}

	for i := 0; i < len(results); i++ {
		if results[i].Ts != uint32(i+1) || results[i].Val != float64(i+1000) {
			t.Fatalf("Unexpected results starting at %d %+v", i, results)
		}
	}
}
```

Reply:
Good catch. I took your test approach and generalized it to a larger range. Catches the underflow from master and the one in that commit.